jax/_src/nn/functions.py

```diff
@@ -765,3 +767,195 @@ def hard_silu(x: ArrayLike) -> Array:
   return x_arr * hard_sigmoid(x_arr)
 
 hard_swish = hard_silu
+
+def _get_large_negative(dtype):
+  dtype_max = jnp.finfo(dtype).max
+  return jnp.asarray(-0.7 * dtype_max, dtype=dtype)
+
+def _get_causal_mask(T, dtype):
+  large_negative_number = _get_large_negative(dtype)
+  col_idx = jnp.tile(jnp.arange(T)[jnp.newaxis, :], [T, 1])
+  row_idx = jnp.tile(jnp.arange(T)[:, jnp.newaxis], [1, T])
+  mask = (row_idx < col_idx).astype(dtype) * large_negative_number
+  mask = mask[jnp.newaxis, jnp.newaxis, :, :]
+  return mask
+
+def _dot_product_attention_xla(
+    query: ArrayLike,
+    key: ArrayLike,
+    value: ArrayLike,
+    bias: ArrayLike | None,
+    mask: ArrayLike | None,
+    is_causal: bool,
+    scale: float,
+    softmax_in_fp32: bool):
+  logits_dtype = jnp.float32 if softmax_in_fp32 else None
+  # Compute the attention logits
+  logits = jnp.einsum('BTNH,BSNH->BNTS', query, key,
+                      preferred_element_type=logits_dtype)
+
+  # Logits scaling
+  scale = jnp.array(scale, dtype=logits.dtype)
+  logits = jnp.multiply(logits, scale)
+
+  # Bias
+  if bias is not None:
+    logits = logits + bias.astype(logits.dtype)
+
+  # Mask
+  if mask is not None:
+    assert mask.dtype == jnp.bool_
+    large_negative_number = _get_large_negative(logits.dtype)
+    padded_logits = jnp.where(mask, logits, large_negative_number)
+  else:
+    padded_logits = logits
+
+  if is_causal:
+    T, S = query.shape[-3], key.shape[-3]
+    assert T == S, f'T should equal to S, but got {T} vs {S}'
+    mask = _get_causal_mask(T, logits.dtype)
+    padded_logits = padded_logits + mask
+
+  # Softmax and it is always carried out in fp32.
+  padded_logits = padded_logits.astype(jnp.float32)
+  probs = jax.nn.softmax(padded_logits, axis=-1).astype(key.dtype)
+
+  # Compute the attention context.
+  encoded = jnp.einsum('BNTS,BSNH->BTNH', probs, value)
+  return encoded, probs
+
+def dot_product_attention(
+    query: ArrayLike,
+    key: ArrayLike,
+    value: ArrayLike,
+    bias: ArrayLike | None = None,
+    mask: ArrayLike | None = None,
+    *,
+    scale: float | None = None,
+    is_causal: bool = False,
+    implementation: str | None = 'xla',
+    softmax_in_fp32: bool = True,
```

Comment:
I think we always compute softmax in fp32 in XLA for bf16 inputs. I prefer the TransformerEngine `float32_logits` option. One ugliness: this is not true for FP64 inputs. One option is to have a `logitits_dtype` that defaults to `None`, which means FP32 for BF16/FP16 inputs, and FP64 for FP64 inputs. What do you think?

Reply:
Or we drop this option for now.

Reply:
Dropped this option. We can add this choice when the fast implementation supports it.